File: native/android/app/src/main/java/com/fantasyac/game/LocalAIEngine.kt
package com.fantasyac.game

import android.content.Context
import org.json.JSONObject

class LocalAIEngine(private val context: Context, private val models: ModelManager) : AutoCloseable {
    companion object { init { System.loadLibrary("fantasyac_llama") } }

    private var handle: Long = 0L
    private var loadedModelId: String? = null

    fun statusJson(): String {
        val active = models.activeModelId()
        return JSONObject()
            .put("available", nativeRuntimeAvailable())
            .put("loaded", handle != 0L && loadedModelId == active)
            .put("modelId", active ?: JSONObject.NULL)
            .put("detail", when {
                active == null -> "로컬 모델이 설치되지 않았습니다."
                handle != 0L && loadedModelId == active -> "Android NDK local narrator ready"
                else -> "모델 설치됨 · 첫 생성 시 로드됩니다."
            }).toString()
    }

    @Synchronized
    private fun ensureActiveModelLoaded(): Boolean {
        val activeId = models.activeModelId() ?: return false
        if (handle != 0L && loadedModelId == activeId) return true
        if (handle != 0L) { nativeFree(handle); handle = 0L; loadedModelId = null }
        val model = models.activeModelFile() ?: return false
        val preset = models.activePresetId()
        val contextSize = when (preset) { "QUALITY" -> 4096; "BATTERY" -> 2048; else -> 3072 }
        val threads = when (preset) { "BATTERY" -> 3; else -> 4 }
        handle = nativeLoadModel(model.absolutePath, contextSize, threads)
        if (handle != 0L) loadedModelId = activeId
        return handle != 0L
    }

    fun generate(requestJson: String): String {
        if (!ensureActiveModelLoaded()) return JSONObject().put("error", "LOCAL_MODEL_NOT_INSTALLED").toString()
        val preset = models.activePresetId()
        val maxTokens = when (preset) { "QUALITY" -> 1600; "BATTERY" -> 800; else -> 1200 }
        val text = nativeGenerate(handle, requestJson, maxTokens, 0.72f, 0.92f)
        return JSONObject().put("text", text).put("provider", "LOCAL").toString()
    }

    @Synchronized
    fun unload() { if (handle != 0L) nativeFree(handle); handle = 0L; loadedModelId = null }
    fun cancel() { if (handle != 0L) nativeCancel(handle) }
    override fun close() = unload()

    private external fun nativeRuntimeAvailable(): Boolean
    private external fun nativeLoadModel(path: String, contextSize: Int, threads: Int): Long
    private external fun nativeGenerate(handle: Long, requestJson: String, maxTokens: Int, temperature: Float, topP: Float): String
    private external fun nativeCancel(handle: Long)
    private external fun nativeFree(handle: Long)
}

File: native/ios/Fantasyac/AI/LocalAIEngine.swift
import Foundation

// Objective-C bridge objects are not annotated Sendable. We keep ownership in a small
// unchecked box only for the synchronous inference hop to a dedicated GCD worker.
// FantasyacLlamaBridge itself serializes generation at the actor level and uses an atomic
// cancellation flag, so this does not permit overlapping generation calls.
private final class FantasyacLlamaBridgeBox: @unchecked Sendable {
    let value: FantasyacLlamaBridge
    init(_ value: FantasyacLlamaBridge) { self.value = value }
}

actor LocalAIEngine {
    static let shared = LocalAIEngine()
    private var modelID: String?
    private var bridge: FantasyacLlamaBridge?
    private var inferenceInFlight = false
    private let models = ModelManager.shared

    func statusDictionary() async -> [String: Any] {
        let active = await models.activeModelId()
        return [
            "available": true,
            "loaded": bridge != nil && modelID == active,
            "busy": inferenceInFlight,
            "modelId": active ?? NSNull(),
            "detail": active == nil ? "로컬 모델이 설치되지 않았습니다." : (bridge != nil && modelID == active ? "iOS local narrator ready" : "모델 설치됨 · 첫 생성 시 로드됩니다.")
        ]
    }

    func generate(requestJson: String) async throws -> String {
        guard !inferenceInFlight else {
            throw NSError(domain: "FantasyacLocalAI", code: 1004, userInfo: [NSLocalizedDescriptionKey: "LOCAL_NARRATOR_BUSY"])
        }
        try await ensureActiveModelLoaded()
        guard let activeBridge = bridge else {
            throw NSError(domain: "FantasyacLocalAI", code: 1002, userInfo: [NSLocalizedDescriptionKey: "LOCAL_MODEL_NOT_LOADED"])
        }
        let preset = await models.activePresetId()
        let maxTokens: Int32 = preset == "QUALITY" ? 1600 : (preset == "BATTERY" ? 800 : 1200)
        inferenceInFlight = true
        defer { inferenceInFlight = false }

        // llama.cpp generation is synchronous. Run it outside the actor executor so cancel()
        // and status calls remain responsive while inference is in progress.
        let bridgeBox = FantasyacLlamaBridgeBox(activeBridge)
        let text: String = try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    let generated = try bridgeBox.value.generateRequestJSON(
                        requestJson,
                        maxTokens: maxTokens,
                        temperature: 0.72,
                        topP: 0.92
                    )
                    continuation.resume(returning: generated)
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
        guard !text.isEmpty else {
            throw NSError(domain: "FantasyacLocalAI", code: 1003, userInfo: [NSLocalizedDescriptionKey: "Local narration failed"])
        }
        return text
    }

    func cancel() { bridge?.cancel() }

    func unload() {
        bridge?.cancel()
        bridge = nil
        modelID = nil
    }

    private func ensureActiveModelLoaded() async throws {
        let active = await models.activeModelId()
        guard let active else {
            throw NSError(domain: "FantasyacLocalAI", code: 1000, userInfo: [NSLocalizedDescriptionKey: "LOCAL_MODEL_NOT_INSTALLED"])
        }
        if bridge != nil && modelID == active { return }
        bridge = nil
        modelID = nil
        guard let model = try await models.activeModelFile() else {
            throw NSError(domain: "FantasyacLocalAI", code: 1000, userInfo: [NSLocalizedDescriptionKey: "LOCAL_MODEL_NOT_INSTALLED"])
        }
        let preset = await models.activePresetId()
        let contextSize: Int32 = preset == "QUALITY" ? 4096 : (preset == "BATTERY" ? 2048 : 3072)
        let native = try FantasyacLlamaBridge(modelPath: model.path, contextSize: contextSize)
        bridge = native
        modelID = active
    }
}
